### PR TITLE
[DO NOT MERGE] Kokkos CUDA Debugging and Sanitizer Integration

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Hey folks,

Over at https://github.com/kokkos/kokkos/issues/783 two years ago @ibaned and @mhoemmen were talking about a magical CUDA debug space which would be like CUDA, but not require a GPU, and be a little less rough than real CUDA development. @crtrott and @dsunder realized that you could cheat a little bit and use the "HBW" space as mockup for your "CUDA" space.

In my view, the goal of an effort like this is to turn the development loop of a multiphysics app Kokkos developer who is just starting a GPU port from

1. Compile (wait N hours)
2. Run and get a segfault
3. Fix that single segfault
4. See (1)

Into a workflow where they hit every mistake at once and get a warning. My implementation is as follows

1) Make HBW into something that just malloc's (this PR)
2) Make a Kokkos Tool which tracks allocations and poisons the HBW space on "Host" kernels, and poisons the "Host" space on "Device" kernels (https://github.com/kokkos/kokkos-tools/pull/63 that PR)

I'm not very adept at Kokkos Core, so I didn't want to make a "CudaDebug" ExecutionSpace, instead my tool just looks at the names of kernels, kernel names prefixed with "Host" are host kernels, those prefixed with "Device" are device ones. A real implementation might include such an ExecutionSpace so that people can properly mark their ExecutionPolicies and such.

The workflow is rather simple, you just compile with -fsanitize=address -fsanitize-recover=address , load the tool, and you get most of the functionality I've described. Optionally, if you add `ASAN_OPTIONS=halt_on_error=0` as an environment variable, you don't die on an error.

TODOS

- [ ] On an error, automatically do some migration (this way consecutive kernels don't warn on the same field)
- [ ] Migrate this functionality from HBWSpace to a new CudaDebugSpace
- [ ] Make a "CudaSimulation" ExecutionSpace

This PR is intended as a discussion for what a solid implementation of this would look like. If somebody was feeling charitable, implementing bullet point three is likely to be the thing I'm most held up by, if somebody knows how to make something which is precisely the "Serial" execution space except it can only see the (HBWSpace/CudaSimulationSpace) memory space, that would be a huge help.

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Separately, I have a version of Clang which will make ASAN print nicer errors, giving people the names of Kokkos::View's they've overrun, and print the Kokkos "Kernel Stack" instead of the function stack you currently get. I can't post that publicly yet, but can probably send it to anybody who wants it from Sandia or collaborators.

**Comment by dalg24** [comment]
Add to whitelist

**Comment by DavidPoliakoff** [comment]
@dalg24 : Is that for me or for the repo maintainers?

**Comment by dalg24** [comment]
This is a command for the GitHub pull request builder on our Jenkins server.  It is me telling the server please test this and all future PRs open by you.

File: core/src/impl/Kokkos_HBWSpace.cpp [hunk 1]
```diff
@@ -115,7 +107,7 @@ void * HBWSpace::allocate( const size_t arg_alloc_size ) const
       // Over-allocate to and round up to guarantee proper alignment.
       size_t size_padded = arg_alloc_size + sizeof(void*) + alignment ;
 
-      void * alloc_ptr = memkind_malloc(MEMKIND_TYPE, size_padded );
+      void * alloc_ptr = malloc( size_padded );
```
Comment (mhoemmen) [inline core/src/impl/Kokkos_HBWSpace.cpp]: Was the idea to take over HBWSpace?  `m_alloc_mech` kinda lies here, right?

Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_HBWSpace.cpp]: @mhoemmen , yeah, in a productized implementation we'd either just use malloc, or steal whatever HostSpace does to use m_alloc_mech. Good catch

Comment (mhoemmen) [reply core/src/impl/Kokkos_HBWSpace.cpp]: Does anyone actually use HBWSpace for its originally intended purpose?

Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_HBWSpace.cpp]: As it is present in develop, HBWSpace is broken, I need to put together a PR for that (it literally won't build as-is). Odds are that'll be a separate PR though

Comment (mhoemmen) [reply core/src/impl/Kokkos_HBWSpace.cpp]: Should we rename it?  `DebugHostSpace`?

Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_HBWSpace.cpp]: @mhoemmen , I was planning to duplicate it. So we'd have HBWSpace (hopefully updated), and DebugHostSpace (perhaps with an alias like "Kokkos::Experimental::SimulatedCudaSpace"), but if we want to get rid of HBWSpace, renaming is a valid option as well

Comment (crtrott) [reply core/src/impl/Kokkos_HBWSpace.cpp]: Yeah we need to fix HBW Space. The problem is we don't really have a test for it running and in production Trinity is running in Cache mode so HBW Space doesn't exist anyway.

Comment (DavidPoliakoff) [reply core/src/impl/Kokkos_HBWSpace.cpp]: @crtrott , then on that other PR to fix HBW (which is a low priority) what I might do is let us at least verify its syntax. One way to do that would be having a (default-off) compile-time switch which makes HBWSpace just malloc/free, and have CI which turns that switch on. That way we could avoid bugs like the current one. I don't know what our answer is for true HBW Space CI, but we could at least make sure that *if* HBW Space functions properly, our semantics for using it are correct.

Comment (mhoemmen) [reply core/src/impl/Kokkos_HBWSpace.cpp]: HBWSpace smells like the first step to a polymorphic allocator (which I would have found useful on occasion).

**Comment by DavidPoliakoff** [comment]
Currently blocked on #2307 (my responsibility), which will allow Memory and Execution spaces which are mechanically identical/similar to other spaces, but which are tracked separately in terms of profiling and access semantics

**Comment by DavidPoliakoff** [comment]
The work for this is being done in #2307

**Pull request closed without merging**